lib/adt.ex

```diff
@@ -24,7 +24,44 @@ defmodule ADT do
       end
     end
     modules = values |> Enum.map(&generate_defmodule/1)
-    Enum.concat([[variants_definition], modules, [variants_reader]])
+    Enum.concat([[variants_definition], modules, [variants_reader, case_macro(values)]])
+  end
+
+  defp case_macro(values) do
+    quote do
+      defmacro case(a, statements) do
+        possible_variants = unquote(values) |> Enum.map(
+          fn {variant, _} ->
+            ADT._shorten(variant)
+          end
+        ) |> Enum.sort
+        given_variants = statements |> Enum.map(fn {k, v} -> { to_string(k), v } end) |> Enum.sort
+        given_variant_names = given_variants |> Enum.map(fn {k, _} -> k end)
+        if possible_variants != given_variant_names do
+          raise ADT._exhaustive_error(possible_variants, given_variant_names)
+        end
+        rules = Enum.flat_map(given_variants, fn {k, v} ->
+          condition = quote do
```

Comment:
Now that this is only one line, why not do away with `condition` altogether and inline it right below?